### PR TITLE
PHP-1355 and PHP-1357: Fix collection and index enumeration for empty first batch

### PR DESCRIPTION
**Pull request by jmikola**
* https://jira.mongodb.org/browse/PHP-1355
 * https://jira.mongodb.org/browse/PHP-1357

### CODE REVIEW

File: log_stream.c [hunk 1]
```diff
@@ -528,10 +528,21 @@ void mongo_log_stream_response_header(mongo_connection *connection, mongo_cursor
...
 
+		/* Command cursors created from a command result document will lack the
+		 * original query document. Allocate a null zval in that case. */
+		if (cursor->query) {
+			query = cursor->query;
+		} else {
+			MAKE_STD_ZVAL(query);
+			ZVAL_NULL(query);
+			free_query = 1;
+		}
```
Comment (derickr) [inline log_stream.c]: I don't think we should allocate something that basically just represents NULL. I've just looked around the code and IMO changing this now (just before 1.6.0) is a bit too dangerous as the code still looks a bit fragile around this.

File: command_cursor.c [hunk 2]
```diff
@@ -130,7 +130,7 @@ int php_mongo_enforce_batch_size_on_command(zval *command, int size TSRMLS_DC)
 /* {{{ MongoCommandCursor iteration helpers */
 int php_mongocommandcursor_is_valid(mongo_command_cursor *cmd_cursor)
 {
-	return cmd_cursor->current != NULL;
+	return cmd_cursor->started_iterating && cmd_cursor->current != NULL;
```
Comment (derickr) [inline command_cursor.c]: Can you please add a nice descriptive comment about this?

File: command_cursor.c [hunk 3]
```diff
@@ -323,10 +323,6 @@ PHP_METHOD(MongoCommandCursor, valid)
 
 	MONGO_CHECK_INITIALIZED(cmd_cursor->zmongoclient, MongoCommandCursor);
 
-	if (!cmd_cursor->started_iterating) {
-		RETURN_FALSE;
-	}
-
```
Comment (derickr) [inline command_cursor.c]: Why did you remove this? I realize that the case might never be hit, but then again, you never know. Leaving it in is IMO safer. Add a comment instead?

File: db.c [hunk 4]
```diff
@@ -696,48 +696,42 @@ static void mongo_db_list_collections_command(zval *this_ptr, zval *options, int
 
 	php_mongo_command_cursor_init_from_document(db->link, cmd_cursor, connection->hash, cursor_env TSRMLS_CC);
 
-	for (
-		php_mongocommandcursor_load_current_element(cmd_cursor TSRMLS_CC);
-		php_mongocommandcursor_is_valid(cmd_cursor);
-		php_mongocommandcursor_advance(cmd_cursor TSRMLS_CC)
+	cmd_cursor->started_iterating = 1;
```
Comment (derickr) [inline db.c]: This looks magical to me. Shouldn't the flag be set from within php_mongo_command_cursor_init_from_document ?

File: db.c [hunk 5]
```diff
@@ -696,48 +696,42 @@ static void mongo_db_list_collections_command(zval *this_ptr, zval *options, int
...
+
+	while (
+		php_mongocommandcursor_advance(cmd_cursor TSRMLS_CC) == SUCCESS &&
+		php_mongocommandcursor_is_valid(cmd_cursor)
 	) {
-		zval **collection_doc;
+		zval **collection_name, **collection_doc = &cmd_cursor->current;
```
Comment (derickr) [inline db.c]: Please split the &cmd_cursor->current initialisation out of the declarations. This looked like a nice WTF :-)

**Comment by derickr** [comment]
LGTM, if you:

 - [ ] Add the comments
 - [ ] Split the declaration and initialisation

File: db.c (hunk 4 above)
Comment (jmikola) [inline db.c]: See: https://jira.mongodb.org/browse/PHP-1362

File: command_cursor.c (hunk 3 above)
Comment (jmikola) [inline command_cursor.c]: As discussed in chat, this is OK in the C function, which we'll call anyway.

See https://jira.mongodb.org/browse/PHP-1363 (related to MongoCursor).

File: log_stream.c (hunk 1 above)
Comment (jmikola) [inline log_stream.c]: As discussed in chat, we'll leave this code as-is (in the PR) but consider refactoring in https://jira.mongodb.org/browse/PHP-1364

File: collection.c [hunk 6]
```diff
@@ -2126,15 +2127,17 @@ void mongo_collection_list_indexes_command(zval *this_ptr, zval *return_value TS
...
+	php_mongocommandcursor_fetch_batch_if_first_is_empty(cmd_cursor TSRMLS_CC);
+	cmd_cursor->started_iterating = 1;
+
 	for (
-		php_mongocommandcursor_load_current_element(cmd_cursor TSRMLS_CC);
-		php_mongocommandcursor_is_valid(cmd_cursor);
-		php_mongocommandcursor_advance(cmd_cursor TSRMLS_CC)
+		iter_status = php_mongocommandcursor_load_current_element(cmd_cursor TSRMLS_CC);
+		iter_status == SUCCESS && php_mongocommandcursor_is_valid(cmd_cursor);
+		iter_status = php_mongocommandcursor_advance(cmd_cursor TSRMLS_CC)
```
Comment (jmikola) [inline collection.c]: @derickr: Please give this another look. I found a bug in the previous implementation of this PR, which caused us to miss the first element when we called `php_mongocommandcursor_advance()` before the first loop iteration. This diff sticks to the original for loop, but adds:

```c
php_mongocommandcursor_fetch_batch_if_first_is_empty(cmd_cursor TSRMLS_CC);
cmd_cursor->started_iterating = 1;
```

...which were the missing pieces from MongoCommandCursor::rewind() when it handles the "init from document" case.

Also, I added error checking of the load/advance return values. The previous code ignored these and simply relied on `php_mongocommandcursor_is_valid()`. While that was fine for most cases, it meant we could miss an exception thrown from the advance function, which might lead us to iterate over the same element multiple times. There might also have been a risk of an infinite loop. If advance continually returned `FAILURE` from the `cmd_cursor->dead` conditional, as it would repeatedly throw exceptions and this function would `add_next_index_zval()` a duplicate value, ad infinitum.

Comment (derickr) [inline collection.c]: These three lines just look *really* odd.

**Comment by derickr** [comment]
LGTM (after testing on 2.8.0rc2, 2.8.0rc3 and 2.8.0rc4)

**Comment by jmikola** [comment]
Tested on 2.6.6 and 2.8 RC2 through RC4.
